src/ucp/wireup/address.c

```diff
@@ -678,27 +735,26 @@ ucs_status_t ucp_address_unpack(ucp_worker_t *worker, const void *buffer,
 
         /* md_index */
         empty_dev    = (*(uint8_t*)ptr) & UCP_ADDRESS_FLAG_EMPTY;
-        ++ptr;
+        ptr = UCS_PTR_TYPE_OFFSET(ptr, uint8_t);
 
         /* device address length */
         dev_addr_len = (*(uint8_t*)ptr) & ~UCP_ADDRESS_FLAG_LAST;
         last_dev     = (*(uint8_t*)ptr) & UCP_ADDRESS_FLAG_LAST;
-        ++ptr;
-
-        ptr += dev_addr_len;
+        ptr = UCS_PTR_TYPE_OFFSET(ptr, uint8_t);
+        ptr = UCS_PTR_BYTE_OFFSET(ptr, dev_addr_len);
 
         last_tl = empty_dev;
         while (!last_tl) {
-            ptr      += sizeof(uint16_t);  /* tl_name_csum */
+            ptr       = UCS_PTR_TYPE_OFFSET(ptr, uint16_t); /* tl_name_csum */
             attr_len  = ucp_address_iface_attr_size(worker);
             flags_ptr = ucp_address_iface_flags_ptr(worker, ptr, attr_len);
-            ptr      += attr_len;
+            ptr       = UCS_PTR_BYTE_OFFSET(ptr, attr_len);
             ptr       = ucp_address_unpack_length(worker, flags_ptr, ptr,
                                                   &iface_addr_len, 0);
-            ptr      += iface_addr_len;
+            ptr       = UCS_PTR_BYTE_OFFSET(ptr, iface_addr_len);
             ptr       = ucp_address_unpack_length(worker, flags_ptr, ptr,
                                                   &ep_addr_len, 1);
-            ptr      += ep_addr_len;
+            ptr       = UCS_PTR_BYTE_OFFSET(ptr, ep_addr_len);
```

Comment:
shouldn't we check of EP unpack flag here? and other flags in line 751?

Reply:
according to packed address format we have to unpack lengths anyway and it defines if corresponding field is present.

Reply:
but you already pass unpack flags and documented they have to be same as pack flags?

Reply:
if unpacked address contains more data then requested this doesn't affect functionality. I tried to keep it as simple as possible in over complicated and widely used part of UCX + don't break something. maybe we need to revise address format to simplify it.

Reply:
> if unpacked address contains more data then requested this doesn't affect functionality
Did you mean packed or unpacked?

I thought we are trying to save as much bytes as possible because private data is limited..

Reply:
> Did you mean packed or unpacked?

unpacked address must contain the same data as packed (this is why I documented requirements for flags) but we can't "to save as much bytes as possible" with current format because some things are packed into the same bytes, e.g. length of smth + UCP_ADDRESS_FLAG_xx. So, address format becomes self defined for some fields but other data can be unpacked (or skipped) only based on flags info.

Reply:
so you mean we can't save bytes by not passing length because it also contains flags?
i don't see any flags together with ep_addr length, what am i missing?

Reply:
about address format in general, WRT ep address and its length see https://github.com/openucx/ucx/pull/3933/files#diff-d92b9d8f5a6ffd97dfe1f3d43660300cR602 and https://github.com/openucx/ucx/blob/56b25b0c2a6b7aefe63ad044415bc6f8f813b596/src/ucp/wireup/address.c#L465